beacon-chain/endtoend/endtoend_test.go

```diff
@@ -0,0 +1,408 @@
+package endtoend
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/big"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/pborman/uuid"
+	contracts "github.com/prysmaticlabs/prysm/contracts/deposit-contract"
+	"github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	prysmKeyStore "github.com/prysmaticlabs/prysm/shared/keystore"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+var log = logrus.WithField("prefix", "e2e")
+
+type End2EndConfig struct {
+	NumValidators  uint64
+	NumBeaconNodes uint64
+}
+
+type BeaconNodeInfo struct {
+	processID   int
+	datadir     string
+	rpcPort     uint64
+	monitorPort uint64
+	grpcPort    uint64
+	multiAddr   string
+}
+
+func TestEndToEnd(t *testing.T) {
+	tmpPath := path.Join("/tmp/e2e/", uuid.NewUUID().String()[:18])
+	os.MkdirAll(tmpPath, os.ModePerm)
+	fmt.Printf("Path for this test is %s\n", tmpPath)
+
+	params.UseDemoBeaconConfig()
+	contractAddr, keystorePath := StartEth1(t, tmpPath)
+	beaconNodes := StartBeaconNodes(t, tmpPath, contractAddr, 1)
+	InitializeValidators(t, tmpPath, contractAddr, keystorePath, beaconNodes, 8)
+
+	beaconLogFile, err := os.Open(path.Join(tmpPath, "beacon-0.log"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := WaitForTextInFile(beaconLogFile, "Chain started within the last epoch"); err != nil {
+		t.Fatal(err)
+	}
+	fmt.Println("Chain has started")
+
+	conn, err := grpc.Dial("127.0.0.1:4000", grpc.WithInsecure())
+	if err != nil {
+		t.Fatalf("fail to dial: %v", err)
+	}
+	time.Sleep(2 * time.Second)
+	beaconClient := eth.NewBeaconChainClient(conn)
+
+	time.Sleep(time.Second * time.Duration(params.BeaconConfig().SecondsPerSlot))
+	currentEpoch := uint64(0)
+	// Run chainstart evaluators outside of the loop.
+	RunChainStartEvaluators(t, beaconClient, currentEpoch)
+
+	scanner := bufio.NewScanner(beaconLogFile)
+	for scanner.Scan() && currentEpoch < 8 {
+		currentLine := scanner.Text()
+		if strings.Contains(currentLine, "Finished applying state transition") {
```

Comment:
Could we use something like `testutil.AssertLogsContain(t, hook, "Finished applying state transition")` instead?

Reply:
The outputs of the beacon chain client are piped into a file, I have to pass through the file line by line that it outputs to make sure the E2E knows when an epoch has passed. Would this let me declare which lines I'm watching?